<a href="https://colab.research.google.com/github/kaixing189/MSSP607/blob/main/WeeklyModules/Week06/Week_06_Participation_Activity.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MSSP607 Week 6 Participation Activity
## Lambda Functions and Student Performance Data

**Student:** Kai Xing

This notebook completes the Week 6 Participation Activity.

### Assignment tasks

1. Review Python lambda functions.
2. Review PEP 312, *Simple Implicit Lambda*.
3. Use Python and Google Colab to create a lambda function that takes the Student Performance Data Set as input and returns another DataFrame.
4. Save the notebook and data file to the MSSP607 GitHub repository.
5. Upload the completed Jupyter Notebook to the Week 6 Participation Activity assignment in Canvas.

## Part 1. Brief Review of Lambda Functions

A lambda expression creates a small anonymous function.

The standard Python syntax is:

```python
lambda parameters: expression
```

For example:

```python
square = lambda x: x ** 2
square(5)
```

returns `25`.

According to the Python documentation, a lambda expression creates a function object. A lambda can contain an expression, but it cannot contain regular statements such as a multi-line `for` loop or `if` statement.

PEP 312 proposed a shorter form of implicit lambda syntax in some situations. However, PEP 312 is currently marked **Deferred**, so this notebook uses the normal Python `lambda` syntax.

References:

- Python Lambda Expressions: https://docs.python.org/3/reference/expressions.html#lambda
- PEP 312 – Simple Implicit Lambda: https://peps.python.org/pep-0312/

## Part 2. Clone the MSSP607 GitHub Repository

This step makes the files in the GitHub repository available inside the current Google Colab runtime.

The repository is public, so a GitHub token is not required for cloning.

Before running this notebook, save `Student Performance Data.csv` in the repository's `data` folder.

The expected GitHub location is:

```text
MSSP607/
└── data/
    └── Student Performance Data.csv
```

In [1]:
from pathlib import Path
import os
import subprocess

REPO_URL = "https://github.com/kaixing189/MSSP607.git"
REPO_DIR = Path("/content/MSSP607")

if REPO_DIR.exists():
    subprocess.run(
        ["git", "-C", str(REPO_DIR), "pull"],
        check=True
    )
    print("Repository already exists. Latest changes were pulled.")
else:
    subprocess.run(
        ["git", "clone", REPO_URL, str(REPO_DIR)],
        check=True
    )
    print("Repository cloned successfully.")

os.chdir(REPO_DIR)
print("Current working directory:", os.getcwd())

Repository cloned successfully.
Current working directory: /content/MSSP607


## Part 3. Load the Student Performance Data Set

This section imports pandas and reads the CSV file into a DataFrame called `students`.

The original data contain 1,000 student records and 10 variables.

In [2]:
import pandas as pd

DATA_PATH = REPO_DIR / "data" / "Student Performance Data.csv"

print("Data file exists:", DATA_PATH.exists())

students = pd.read_csv(DATA_PATH)

print("DataFrame shape:", students.shape)
students.head()

Data file exists: True
DataFrame shape: (1000, 10)


,gender,race/ethnicity,parental level of education,lunch,math score,reading score,writing score,total score,test preparation course,admission prospect
0,female,group B,bachelor's degree,standard,72,72,74,218,none,medium
1,female,group C,some college,standard,69,90,88,247,completed,high
2,female,group B,master's degree,standard,90,95,93,278,none,high
3,male,group A,associate's degree,free/reduced,47,57,44,148,none,low
4,male,group C,some college,standard,76,78,75,229,none,high


## Part 4. Inspect the Data

Before applying the lambda function, it is useful to check the column names and data types.

The three academic score variables are:

- `math score`
- `reading score`
- `writing score`

These three variables will be used to calculate each student's average score.

In [3]:
print("Columns:")
print(students.columns.tolist())

print("\nData types:")
print(students.dtypes)

Columns:
['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'math score', 'reading score', 'writing score', 'total score', 'test preparation course', 'admission prospect']

Data types:
gender                         object
race/ethnicity                 object
parental level of education    object
lunch                          object
math score                      int64
reading score                   int64
writing score                   int64
total score                     int64
test preparation course        object
admission prospect             object
dtype: object


The next cell checks for missing values.

In [4]:
students.isna().sum()

,0
gender,0
race/ethnicity,0
parental level of education,0
lunch,0
math score,0
reading score,0
writing score,0
total score,0
test preparation course,0
admission prospect,0


## Part 5. Create a Lambda Function That Returns Another DataFrame

### Main assignment requirement

The following lambda function takes a DataFrame named `data` as its input.

It uses `assign()` to return a **new DataFrame** with an additional column called `average score`.

The average is calculated using the student's math, reading, and writing scores.

The original `students` DataFrame is not overwritten.

In [6]:
create_score_dataframe = lambda data: data.assign(
    **{
        "average score": lambda x: x[
            ["math score", "reading score", "writing score"]
        ].mean(axis=1)
    }
)

## Part 6. Apply the Lambda Function

The lambda function is now applied to the complete Student Performance Data Set.

The result is stored in a new DataFrame called `student_scores`.

The original DataFrame has 10 columns. The new DataFrame has 11 columns because it includes the new `average score` variable.

In [7]:
student_scores = create_score_dataframe(students)

print("Original DataFrame shape:", students.shape)
print("New DataFrame shape:", student_scores.shape)

student_scores.head()

Original DataFrame shape: (1000, 10)
New DataFrame shape: (1000, 11)


,gender,race/ethnicity,parental level of education,lunch,math score,reading score,writing score,total score,test preparation course,admission prospect,average score
0,female,group B,bachelor's degree,standard,72,72,74,218,none,medium,72.666667
1,female,group C,some college,standard,69,90,88,247,completed,high,82.333333
2,female,group B,master's degree,standard,90,95,93,278,none,high,92.666667
3,male,group A,associate's degree,free/reduced,47,57,44,148,none,low,49.333333
4,male,group C,some college,standard,76,78,75,229,none,high,76.333333


### Interpretation

The lambda function successfully takes the original Student Performance DataFrame as input and returns another DataFrame.

The returned DataFrame contains all original student information plus a new `average score` column. This column summarizes each student's performance across the math, reading, and writing tests.

For example, if a student has scores of 72 in math, 72 in reading, and 74 in writing, the average score is approximately 72.67.

## Part 7. Verify That a New DataFrame Was Returned

This step confirms two things:

1. The original DataFrame still has 10 columns.
2. The returned DataFrame has the new `average score` column.

This demonstrates that the lambda function returned a separate DataFrame rather than changing the original object directly.

In [8]:
print("Original columns:", len(students.columns))
print("New columns:", len(student_scores.columns))

print("\nIs 'average score' in original DataFrame?")
print("average score" in students.columns)

print("\nIs 'average score' in new DataFrame?")
print("average score" in student_scores.columns)

Original columns: 10
New columns: 11

Is 'average score' in original DataFrame?
False

Is 'average score' in new DataFrame?
True


## Part 8. View Summary Statistics for the New Variable

This section is not required to define the lambda function, but it provides a useful check of the new DataFrame.

The summary shows the distribution of the newly calculated `average score`.

In [9]:
student_scores["average score"].describe()

,average score
count,1000.000000
mean,67.770667
std,14.257326
min,9.000000
25%,58.333333
50%,68.333333
75%,77.666667
max,100.000000


## Part 9. Optional: View Students with the Highest Average Scores

This step sorts the returned DataFrame by `average score` and displays the first ten rows.

It demonstrates that the new DataFrame can be used for additional analysis after the lambda transformation.

In [10]:
student_scores.sort_values(
    by="average score",
    ascending=False
).head(10)

,gender,race/ethnicity,parental level of education,lunch,math score,reading score,writing score,total score,test preparation course,admission prospect,average score
916,male,group E,bachelor's degree,standard,100,100,100,300,completed,high,100.000000
962,female,group E,associate's degree,standard,100,100,100,300,none,high,100.000000
458,female,group E,bachelor's degree,standard,100,100,100,300,none,high,100.000000
114,female,group E,bachelor's degree,standard,99,100,100,299,completed,high,99.666667
712,female,group D,some college,standard,98,100,99,297,none,high,99.000000
179,female,group D,some high school,standard,97,100,100,297,completed,high,99.000000
165,female,group C,bachelor's degree,standard,96,100,100,296,completed,high,98.666667
625,male,group D,some college,standard,100,97,99,296,completed,high,98.666667
685,female,group E,master's degree,standard,94,99,100,293,completed,high,97.666667
903,female,group D,bachelor's degree,free/reduced,93,100,100,293,completed,high,97.666667


## Part 10. Final Verification

The following checks confirm that the activity was completed correctly.

In [11]:
assert students.shape == (1000, 10)
assert student_scores.shape == (1000, 11)
assert "average score" not in students.columns
assert "average score" in student_scores.columns

print("All checks passed.")
print("The lambda function successfully returned a new DataFrame.")

All checks passed.
The lambda function successfully returned a new DataFrame.


## Part 11. Save the Notebook and Data to GitHub

After running all cells successfully:

1. In Colab, select **File → Save a copy in GitHub**.
2. Select the repository `kaixing189/MSSP607`.
3. Use the following file path:

```text
WeeklyModules/Week06/Week_06_Participation_Activity.ipynb
```

4. A suggested commit message is:

```text
Complete Week 6 participation activity
```

5. Make sure the data file is also stored in:

```text
data/Student Performance Data.csv
```

6. Confirm on the GitHub website that both the notebook and data file are visible.
7. Confirm that the instructor and TAs have collaborator access if required by the course.

## Part 12. Canvas Submission Checklist

Before submitting:

- [ ] Run **Runtime → Run all** in Google Colab.
- [ ] Confirm that every code cell runs without an error.
- [ ] Keep the code outputs visible.
- [ ] Confirm the notebook is saved in `WeeklyModules/Week06/` on GitHub.
- [ ] Confirm `Student Performance Data.csv` is saved in the GitHub `data/` folder.
- [ ] Upload the final `.ipynb` file to the Week 6 Participation Activity in Canvas.
- [ ] Check the Canvas submission page to confirm the upload was successful.